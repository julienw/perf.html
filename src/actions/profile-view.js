/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  selectorsForThread,
  selectedThreadSelectors,
} from '../reducers/profile-view';
import {
  getImplementationFilter,
  getSelectedThreadIndex,
} from '../reducers/url-state';

import type { ProfileSelection, ImplementationFilter } from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoMarkersTable } from '../types/profile';
import type { CallNodePath } from '../types/profile-derived';
import type { Transform } from '../types/transforms';
/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedCallNode(
  threadIndex: ThreadIndex,
  selectedCallNodePath: CallNodePath
): Action {
  return {
    type: 'CHANGE_SELECTED_CALL_NODE',
    selectedCallNodePath,
    threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(
      ({
        type: 'HIDE_THREAD',
        threadIndex,
        threadOrder,
        hiddenThreads,
      }: Action)
    );
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedCallNodes(
  threadIndex: ThreadIndex,
  expandedCallNodePaths: Array<CallNodePath>
): Action {
  return {
    type: 'CHANGE_EXPANDED_CALL_NODES',
    threadIndex,
    expandedCallNodePaths,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker,
    threadIndex,
  };
}

export function changeMarkersSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_MARKER_SEARCH_STRING',
    searchString,
  };
}

export function changeImplementationFilter(
  implementation: ImplementationFilter
): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousImplementation = getImplementationFilter(getState());
    const threadIndex = getSelectedThreadIndex(getState());
    const transformedThread = selectedThreadSelectors.getRangeAndTransformFilteredThread(
      getState()
    );

    dispatch({
      type: 'CHANGE_IMPLEMENTATION_FILTER',
      implementation,
      threadIndex,
      transformedThread,
      previousImplementation,
    });
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(
  hidePlatformDetails: boolean
): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start,
    end,
  };
}

export function addRangeFilterAndUnsetSelection(
  start: number,
  end: number
): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(
      updateProfileSelection({ hasSelection: false, isModifying: false })
    );
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(
  firstRemovedFilterIndex: number
): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(
      updateProfileSelection({ hasSelection: false, isModifying: false })
    );
  };
}

export function addTransformToStack(
  threadIndex: ThreadIndex,
  transform: Transform
): ThunkAction<void> {
  return (dispatch, getState) => {
    const transformedThread = selectorsForThread(
      threadIndex
    ).getRangeAndTransformFilteredThread(getState());

    dispatch({
      type: 'ADD_TRANSFORM_TO_STACK',
      threadIndex,
      transform,
      transformedThread,
    });
  };
}

export function popTransformsFromStack(
  threadIndex: ThreadIndex,
  firstRemovedFilterIndex: number
): Action {
  return {
    type: 'POP_TRANSFORMS_FROM_STACK',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
