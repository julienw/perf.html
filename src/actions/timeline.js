/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';
import type { GetLabel } from '../profile-logic/labeling-strategies';
import type { GetCategory } from '../profile-logic/color-categories';

export function changeFlameChartColorStrategy(
  getCategory: GetCategory
): Action {
  return {
    type: 'CHANGE_FLAME_CHART_COLOR_STRATEGY',
    getCategory,
  };
}

export function changeFlameChartLabelingStrategy(getLabel: GetLabel): Action {
  return {
    type: 'CHANGE_FLAME_CHART_LABELING_STRATEGY',
    getLabel,
  };
}

export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}
