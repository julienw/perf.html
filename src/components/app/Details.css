.Details {
  flex: auto;

  display: flex;
  flex-direction: column;
}
