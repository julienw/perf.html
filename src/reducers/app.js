/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import { getSelectedTab } from './url-state';
import { tabSlugs } from '../app-logic/tabs-handling';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { Action } from '../types/store';
import type {
  State,
  AppState,
  AppViewState,
  IsSidebarOpenPerPanelState,
  Reducer,
} from '../types/reducers';

function view(
  state: AppViewState = { phase: 'INITIALIZING' },
  action: Action
): AppViewState {
  if (state.phase === 'DATA_LOADED') {
    // Let's not come back at another phase if we're already displaying a profile
    return state;
  }

  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_URL':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_ADDON':
      return {
        phase: 'INITIALIZING',
        additionalData: {
          message: action.error.message,
          attempt: action.error.attempt,
        },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_URL':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'WAITING_FOR_PROFILE_FROM_ADDON':
      return { phase: 'INITIALIZING' };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_ZIP_FILE':
    case 'VIEW_PROFILE':
      return { phase: 'DATA_LOADED' };
    default:
      return state;
  }
}

function isUrlSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'URL_SETUP_DONE':
      return true;
    default:
      return state;
  }
}

function hasZoomedViaMousewheel(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'HAS_ZOOMED_VIA_MOUSEWHEEL': {
      return true;
    }
    default:
      return state;
  }
}

function isSidebarOpenPerPanel(
  state: IsSidebarOpenPerPanelState,
  action: Action
): IsSidebarOpenPerPanelState {
  if (state === undefined) {
    state = {};
    tabSlugs.forEach(tabSlug => (state[tabSlug] = false));
  }

  switch (action.type) {
    case 'CHANGE_SIDEBAR_OPEN_STATE': {
      const { tab, isOpen } = action;
      // Due to how this action will be dispatched we'll always have the value
      // changed so we don't need the performance optimization of checking the
      // stored value against the new value.
      return {
        ...state,
        [tab]: isOpen,
      };
    }
    default:
      return state;
  }
}

/**
 * The panels that make up the timeline, details view, and sidebar can all change
 * their sizes depending on the state that is fed to them. In order to control
 * the invalidations of this sizing information, provide a "generation" value that
 * increases monotonically for any change that potentially changes the sizing of
 * any of the panels. This provides a mechanism for subscribing components to
 * deterministically update their sizing correctly.
 */
function panelLayoutGeneration(state: number = 0, action: Action): number {
  switch (action.type) {
    case 'INCREMENT_PANEL_LAYOUT_GENERATION':
    // Sidebar: (fallthrough)
    case 'CHANGE_SIDEBAR_OPEN_STATE':
    // Timeline: (fallthrough)
    case 'HIDE_GLOBAL_TRACK':
    case 'SHOW_GLOBAL_TRACK':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'HIDE_LOCAL_TRACK':
    case 'SHOW_LOCAL_TRACK':
    case 'ISOLATE_LOCAL_TRACK':
    // Committed range changes: (fallthrough)
    case 'COMMIT_RANGE':
    case 'POP_COMMITTED_RANGES':
      return state + 1;
    default:
      return state;
  }
}

/**
 * Clicking on tracks can switch between different tabs. This piece of state holds
 * on to the last relevant thread-based tab that was viewed. This makes the UX nicer
 * for when a user clicks on a Network track, and gets taken to the Network
 * panel, then clicks on a thread or process track. With this state we can smoothly
 * transition them back to the panel they were using.
 */
function lastVisibleThreadTabSlug(
  state: TabSlug | null = null,
  action: Action
): TabSlug | null {
  switch (action.type) {
    case 'SELECT_TRACK':
    case 'CHANGE_SELECTED_TAB':
      if (action.selectedTab !== 'network-chart') {
        return action.selectedTab;
      }
      return state;
    default:
      return state;
  }
}

const appStateReducer: Reducer<AppState> = combineReducers({
  view,
  isUrlSetupDone,
  hasZoomedViaMousewheel,
  isSidebarOpenPerPanel,
  panelLayoutGeneration,
  lastVisibleThreadTabSlug,
});

export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsUrlSetupDone = (state: State): boolean =>
  getApp(state).isUrlSetupDone;
export const getHasZoomedViaMousewheel = (state: State): boolean => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen = (state: State): boolean =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration = (state: State) =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug = (state: State) =>
  getApp(state).lastVisibleThreadTabSlug;
