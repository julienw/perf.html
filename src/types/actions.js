/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { CallTree } from '../profile-logic/call-tree';
import type {
  Profile,
  Thread,
  ThreadIndex,
  IndexIntoMarkersTable,
  IndexIntoFuncTable,
} from './profile';
import type { IndexIntoCallNodeTable, CallNodeTable } from './profile-derived';
import type { GetLabel } from '../profile-logic/labeling-strategies';
import type { GetCategory } from '../profile-logic/color-categories';
import type { TemporaryError } from '../utils/errors';
import type { Transform } from './transforms';

export type DataSource =
  | 'none'
  | 'from-file'
  | 'from-addon'
  | 'local'
  | 'public'
  | 'from-url';
export type ProfileSelection =
  | { hasSelection: false, isModifying: false }
  | {
      hasSelection: true,
      isModifying: boolean,
      selectionStart: number,
      selectionEnd: number,
    };
export type FuncToFuncMap = Map<IndexIntoFuncTable, IndexIntoFuncTable>;
export type FunctionsUpdatePerThread = {
  [id: ThreadIndex]: {
    oldFuncToNewFuncMap: FuncToFuncMap,
    funcIndices: IndexIntoFuncTable[],
    funcNames: string[],
  },
};

export type RequestedLib = { debugName: string, breakpadId: string };
export type ImplementationFilter = 'combined' | 'js' | 'cpp';
export type TabSlug =
  | 'calltree'
  | 'stack-chart'
  | 'marker-chart'
  | 'marker-table'
  | 'flame-graph';

type ProfileAction =
  | { type: 'ROUTE_NOT_FOUND', url: string }
  | { type: 'CHANGE_THREAD_ORDER', threadOrder: ThreadIndex[] }
  | {
      type: 'HIDE_THREAD',
      threadIndex: ThreadIndex,
      hiddenThreads: ThreadIndex[],
      threadOrder: ThreadIndex[],
    }
  | { type: 'SHOW_THREAD', threadIndex: ThreadIndex }
  | {
      type: 'ISOLATE_THREAD',
      hiddenThreadIndexes: ThreadIndex[],
      isolatedThreadIndex: ThreadIndex,
    }
  | {
      type: 'ASSIGN_TASK_TRACER_NAMES',
      addressIndices: number[],
      symbolNames: string[],
    }
  | {
      type: 'CHANGE_SELECTED_CALL_NODE',
      threadIndex: ThreadIndex,
      selectedCallNodeIndex: IndexIntoCallNodeTable | null,
      callNodeAncestors: IndexIntoCallNodeTable[],
    }
  | {
      type: 'FOCUS_CALL_TREE',
    }
  | {
      type: 'CHANGE_EXPANDED_CALL_NODES',
      threadIndex: ThreadIndex,
      expandedCallNodeIndexes: Set<IndexIntoCallNodeTable>,
    }
  | {
      type: 'CHANGE_SELECTED_MARKER',
      threadIndex: ThreadIndex,
      selectedMarker: IndexIntoMarkersTable | -1,
    }
  | { type: 'UPDATE_PROFILE_SELECTION', selection: ProfileSelection }
  | { type: 'CHANGE_TAB_ORDER', tabOrder: number[] };

type ReceiveProfileAction =
  | {
      type: 'COALESCED_FUNCTIONS_UPDATE',
      functionsUpdatePerThread: FunctionsUpdatePerThread,
    }
  | { type: 'DONE_SYMBOLICATING' }
  | { type: 'ERROR_RECEIVING_PROFILE_FROM_FILE', error: Error }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_ADDON',
      error: TemporaryError,
    }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_ADDON', error: Error }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_STORE',
      error: TemporaryError,
    }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_URL',
      error: TemporaryError,
    }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_STORE', error: Error }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_URL', error: Error }
  | { type: 'RECEIVE_PROFILE_FROM_ADDON', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_FILE', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_STORE', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_URL', profile: Profile }
  | { type: 'REQUESTING_SYMBOL_TABLE', requestedLib: RequestedLib }
  | { type: 'RECEIVED_SYMBOL_TABLE_REPLY', requestedLib: RequestedLib }
  | { type: 'START_SYMBOLICATING' }
  | { type: 'WAITING_FOR_PROFILE_FROM_ADDON' }
  | { type: 'WAITING_FOR_PROFILE_FROM_STORE' }
  | { type: 'WAITING_FOR_PROFILE_FROM_URL' };

type StackChartAction =
  | { type: 'CHANGE_STACK_CHART_COLOR_STRATEGY', getCategory: GetCategory }
  | { type: 'CHANGE_STACK_CHART_LABELING_STRATEGY', getLabel: GetLabel }
  | { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };

type UrlEnhancerAction =
  | { type: '@@urlenhancer/urlSetupDone' }
  | { type: '@@urlenhancer/updateUrlState', urlState: any };

type UrlStateAction =
  | { type: 'WAITING_FOR_PROFILE_FROM_FILE' }
  | { type: 'PROFILE_PUBLISHED', hash: string }
  | { type: 'CHANGE_SELECTED_TAB', selectedTab: TabSlug }
  | { type: 'ADD_RANGE_FILTER', start: number, end: number }
  | { type: 'POP_RANGE_FILTERS', firstRemovedFilterIndex: number }
  | { type: 'CHANGE_SELECTED_THREAD', selectedThread: ThreadIndex }
  | { type: 'CHANGE_RIGHT_CLICKED_THREAD', selectedThread: ThreadIndex }
  | { type: 'CHANGE_CALL_TREE_SEARCH_STRING', searchString: string }
  | {
      type: 'ADD_TRANSFORM_TO_STACK',
      threadIndex: ThreadIndex,
      transform: Transform,
      transformedThread: Thread,
    }
  | {
      type: 'POP_TRANSFORMS_FROM_STACK',
      threadIndex: ThreadIndex,
      firstRemovedFilterIndex: number,
    }
  | {
      type: 'CHANGE_IMPLEMENTATION_FILTER',
      implementation: ImplementationFilter,
      threadIndex: ThreadIndex,
      transformedThread: Thread,
      previousImplementation: ImplementationFilter,
      implementation: ImplementationFilter,
    }
  | {|
      type: 'CHANGE_INVERT_CALLSTACK',
      invertCallstack: boolean,
      callTree: CallTree,
      callNodeTable: CallNodeTable,
      selectedThreadIndex: ThreadIndex,
    |}
  | { type: 'CHANGE_HIDE_PLATFORM_DETAILS', hidePlatformDetails: boolean }
  | { type: 'CHANGE_MARKER_SEARCH_STRING', searchString: string };

type IconsAction =
  | { type: 'ICON_HAS_LOADED', icon: string }
  | { type: 'ICON_IN_ERROR', icon: string };

export type Action =
  | ProfileAction
  | ReceiveProfileAction
  | StackChartAction
  | UrlEnhancerAction
  | UrlStateAction
  | IconsAction;
