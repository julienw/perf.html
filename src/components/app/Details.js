/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import explicitConnect from '../../utils/connect';
import TabBar from './TabBar';
import ProfileCallTreeView from '../calltree/ProfileCallTreeView';
import MarkerTable from '../marker-table';
import StackChart from '../stack-chart/';
import MarkerChart from '../marker-chart/';
import FlameGraph from '../flame-graph/';
import { changeSelectedTab, changeTabOrder } from '../../actions/app';
import { getTabOrder } from '../../reducers/profile-view';
import { getSelectedTab } from '../../reducers/url-state';
import CallNodeContextMenu from '../shared/CallNodeContextMenu';
import MarkerTableContextMenu from '../marker-table/ContextMenu';
import ProfileThreadHeaderContextMenu from '../header/ProfileThreadHeaderContextMenu';
import { toValidTabSlug } from '../../utils/flow';

import type { Tab } from './TabBar';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';
import type { TabSlug } from '../../types/actions';

require('./Details.css');

type StateProps = {|
  +tabOrder: number[],
  +selectedTab: TabSlug,
|};

type DispatchProps = {|
  +changeSelectedTab: typeof changeSelectedTab,
  +changeTabOrder: typeof changeTabOrder,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class ProfileViewer extends PureComponent<Props> {
  // If updating this list, make sure and update the tabOrder reducer with another index.
  _tabs: Tab[] = [
    {
      name: 'calltree',
      title: 'Call Tree',
    },
    {
      name: 'flame-graph',
      title: 'Flame Graph',
    },
    {
      name: 'stack-chart',
      title: 'Stack Chart',
    },
    {
      name: 'marker-chart',
      title: 'Marker Chart',
    },
    {
      name: 'marker-table',
      title: 'Marker Table',
    },
    {
      name: 'network-chart',
      title: 'Network',
    },
  ];

  _onSelectTab = (selectedTab: string) => {
    const { changeSelectedTab } = this.props;
    const tabSlug = toValidTabSlug(selectedTab);
    if (!tabSlug) {
      throw new Error('Attempted to change to a tab that does not exist.');
    }
    changeSelectedTab(tabSlug);
  };

  render() {
    const { tabOrder, changeTabOrder, selectedTab } = this.props;
    return (
      <div className="Details">
        <TabBar
          tabs={this._tabs}
          selectedTabName={selectedTab}
          tabOrder={tabOrder}
          onSelectTab={this._onSelectTab}
          onChangeTabOrder={changeTabOrder}
        />
        {
          {
            calltree: <ProfileCallTreeView />,
            'marker-table': <MarkerTable />,
            'stack-chart': <StackChart />,
            'marker-chart': <MarkerChart />,
            'network-chart': <MarkerChart />,
            'flame-graph': <FlameGraph />,
          }[selectedTab]
        }
        <CallNodeContextMenu />
        <MarkerTableContextMenu />
        <ProfileThreadHeaderContextMenu />
      </div>
    );
  }
}

const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
  mapStateToProps: state => ({
    tabOrder: getTabOrder(state),
    selectedTab: getSelectedTab(state),
  }),
  mapDispatchToProps: {
    changeSelectedTab,
    changeTabOrder,
  },
  component: ProfileViewer,
};

export default explicitConnect(options);
