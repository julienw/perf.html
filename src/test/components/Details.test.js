/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { shallowWithStore } from '../fixtures/enzyme';

import Details from '../../components/app/Details';
import { changeSelectedTab } from '../../actions/app';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

import type { TabSlug } from '../../types/actions';

describe('app/DetailsContainer', function() {
  const { profile } = getProfileFromTextSamples(`
    A A A
    B B B
    C C H
    D F I
    E E
  `);

  const tabSlugs: TabSlug[] = [
    'stack-chart',
    'marker-chart',
    'flame-graph',
    'marker-table',
    'calltree', // 'calltree' not first on purpose
  ];

  it('renders an initial view with the right panel', () => {
    const store = storeWithProfile(profile);
    // dive() will shallow-render the wrapped component
    const view = shallowWithStore(<Details />, store);
    expect(view.dive()).toMatchSnapshot();
  });

  tabSlugs.forEach((tabSlug: TabSlug) => {
    it(`renders an initial view with the right panel for tab ${tabSlug}`, () => {
      const store = storeWithProfile(profile);
      store.dispatch(changeSelectedTab(tabSlug));

      const view = shallowWithStore(<Details />, store);
      expect(view.dive()).toMatchSnapshot();
    });
  });
});
