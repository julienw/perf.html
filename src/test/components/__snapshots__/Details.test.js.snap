// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/DetailsContainer renders an initial view with the right panel 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="calltree"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;

exports[`app/DetailsContainer renders an initial view with the right panel for tab calltree 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="calltree"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;

exports[`app/DetailsContainer renders an initial view with the right panel for tab flame-graph 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="flame-graph"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <FlameGraphView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;

exports[`app/DetailsContainer renders an initial view with the right panel for tab marker-chart 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="marker-chart"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <Connect(MarkerChart) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;

exports[`app/DetailsContainer renders an initial view with the right panel for tab marker-table 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="marker-table"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <Connect(MarkerTable) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;

exports[`app/DetailsContainer renders an initial view with the right panel for tab stack-chart 1`] = `
<div
  className="Details"
>
  <TabBar
    onChangeTabOrder={[Function]}
    onSelectTab={[Function]}
    selectedTabName="stack-chart"
    tabOrder={
      Array [
        0,
        1,
        2,
        3,
        4,
        5,
      ]
    }
    tabs={
      Array [
        Object {
          "name": "calltree",
          "title": "Call Tree",
        },
        Object {
          "name": "flame-graph",
          "title": "Flame Graph",
        },
        Object {
          "name": "stack-chart",
          "title": "Stack Chart",
        },
        Object {
          "name": "marker-chart",
          "title": "Marker Chart",
        },
        Object {
          "name": "marker-table",
          "title": "Marker Table",
        },
        Object {
          "name": "network-chart",
          "title": "Network",
        },
      ]
    }
  />
  <Connect(StackChartGraph) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(ProfileThreadHeaderContextMenu) />
</div>
`;
